### PR TITLE
[WIP] refactored authorization event/event listeners

### PR DESCRIPTION
**Pull request by MichaelKubovic**

### CODE REVIEW

**Comment by ajgarlag** [comment]
@MichaelKubovic Can you resolve conflict in order to review your PR?

Thanks!

**Comment by MichaelKubovic** [comment]
Sorry for the delay. Done, please have a look. If https://github.com/trikoder/oauth2-bundle/pull/18 gets merged anytime soon, I could move the PR there. But for now I'd appreciate your feedback as it is.

File: Event/Listener/AuthorizationRequestAuthenticationListener.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+     * @param RequestStack $requestStack
+     * @param UrlGeneratorInterface $urlGenerator
+     * @param string $loginRoute
+     */
+    public function __construct(
+        AuthorizationCheckerInterface $authorizationChecker,
+        SessionInterface $session,
+        RequestStack $requestStack,
+        UrlGeneratorInterface $urlGenerator,
+        string $loginRoute = 'app_login'
```
Comment (ajgarlag) [inline Event/Listener/AuthorizationRequestAuthenticationListener.php]: This route should be easily defined from bundle configuration.

File: Event/Listener/AuthorizationRequestAuthenticationListener.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+     * @param AuthorizationRequestResolveEvent $event
+     */
+    public function onAuthorizationRequest(AuthorizationRequestResolveEvent $event): void
+    {
+        if (null === $request = $this->requestStack->getMasterRequest()) {
+            throw new \RuntimeException('Authentication listener depends on the request context');
+        }
+
+        if (!$this->authorizationChecker->isGranted('IS_AUTHENTICATED_REMEMBERED')) {
+            $this->saveTargetPath($this->session, 'main', $request->getUri());
```
Comment (ajgarlag) [inline Event/Listener/AuthorizationRequestAuthenticationListener.php]: The `$providerKey` argument should not be hardcoded to `main`, but derived from current `HttpFoundation\Request`

File: Event/Listener/AuthorizationRequestAuthenticationListener.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    /**
+     * @param AuthorizationRequestResolveEvent $event
+     */
+    public function onAuthorizationRequest(AuthorizationRequestResolveEvent $event): void
+    {
+        if (null === $request = $this->requestStack->getMasterRequest()) {
+            throw new \RuntimeException('Authentication listener depends on the request context');
+        }
+
+        if (!$this->authorizationChecker->isGranted('IS_AUTHENTICATED_REMEMBERED')) {
```
Comment (ajgarlag) [inline Event/Listener/AuthorizationRequestAuthenticationListener.php:78]: What to do if I want to require `IS_AUTHENTICATED_FULLY`?

File: Event/AuthorizationRequestResolveEvent.php [hunk 4]
```diff
@@ -7,64 +7,32 @@
...
-    public const AUTHORIZATION_APPROVED = true;
-    public const AUTHORIZATION_DENIED = false;
-
     /**
      * @var AuthorizationRequest
      */
     private $authorizationRequest;
 
     /**
-     * @var ?ResponseInterface
```
Comment (ajgarlag) [inline Event/AuthorizationRequestResolveEvent.php]: I prefer to use the interface instead of the concrete implementation when available

File: Event/AuthorizationRequestResolveEvent.php [hunk 5]
```diff
@@ -7,64 +7,32 @@
...
-        return $this->authorizationResolution;
-    }
-
-    public function resolveAuthorization(bool $authorizationResolution): void
-    {
-        $this->authorizationResolution = $authorizationResolution;
-        $this->response = null;
-    }
-
-    public function hasResponse(): bool
```
Comment (ajgarlag) [inline Event/AuthorizationRequestResolveEvent.php]: Do not remove the response related method to introduce them again

File: Model/AuthorizationDecision/UserConsentDecisionStrategy.php [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+     * @param UriSigner $uriSigner
+     * @param RequestStack $requestStack
+     * @param UrlGeneratorInterface $urlGenerator
+     * @param string $consentApprovalRoute
+     */
+    public function __construct(
+        UriSigner $uriSigner,
+        RequestStack $requestStack,
+        UrlGeneratorInterface $urlGenerator,
+        string $consentApprovalRoute = 'oauth2_consent'
```
Comment (ajgarlag) [inline Model/AuthorizationDecision/UserConsentDecisionStrategy.php]: This route should be overridden from bundle configuration.

File: DependencyInjection/Configuration.php [hunk 7]
```diff
@@ -58,6 +58,11 @@ private function createAuthorizationServerNode(): NodeDefinition
                     ->cannotBeEmpty()
                     ->defaultValue('PT10M')
                 ->end()
+                ->enumNode('authorization_strategy')
+                    ->info("What strategy should be used to authorize user.\nAvailable options are: user_consent - let user authorize the client request, always_allow - silently authorizes all client requests")
+                    ->values(['user_consent', 'always_allow'])
```
Comment (ajgarlag) [inline DependencyInjection/Configuration.php]: An user cannot implement a custom strategy

**Review by ajgarlag — COMMENTED** [review]
The main difference I see is that in my implementation the `AuthorizationRequest` instance is never mutated from the event listeners, but with your implementation, only two modification are allowed (setting the user and approving the request). I think this can be proposed to the upstream repository once https://github.com/trikoder/oauth2-bundle/pull/18 gets merged. Which I want to avoid is to expose the `AuthorizationRequest` instance from the event.

Another difference is that in my implementation, when an event listener decide about the `AuthorizationRequest` (allowing or denying it), the response is cleared so the user is redirected only if there is no decision.

I do not like the `authorization_strategy` related changes for different reasons:
  - This functionality can be implemented without introducing another level of indirection. Simply configure the appropriate event listeners.
  - To implement a new strategy, the bundle configuration should be modified, so it is difficult to extend it. If we only use event listeners, this problem is avoided.
  - If I choose the `user_consent` strategy, I'd like to set the `$consentApprovalRoute` from the bundle configuration, so the bundle config gets more complicated.

My thought was to left to the bundle user to decide how to implement the authorization process, but I think it can be convenient to provide an `AlwaysAllowAuthorizationRequestListener`, that could be easily enabled from a bundle configuration key, so we provide a minimal working implementation.

**Comment by MichaelKubovic** [comment]
I should have stated the motivation behind this PR first:
Firstly, moving all `AuthorizationRequest` mutations to listeners has many benefits:
- it's makes the Authorization controller a mere dispatcher with the focus on the final state of the event
- having a well-defined sequence of listeners increases readability (what mutations happens when)
- it allows anybody to use CompilerPass to change only pieces they are interested in rather then overriding the whole controller
- plays nice with the plug&play listeners (I had the requirement to add the authentication listener myself)
I wanted to make a clear distinction between the internal design (listeners acting on the request) and a very focused extension points (how the authorization is done)

> The main difference I see is that in my implementation the `AuthorizationRequest` instance is never mutated from the event listeners, but with your implementation, only two modification are allowed (setting the user and approving the request). I think this can be proposed to the upstream repository once [trikoder#18](https://github.com/trikoder/oauth2-bundle/pull/18) gets merged. Which I want to avoid is to expose the `AuthorizationRequest` instance from the event.

Agree, encapsulation around `AuthorizationRequest` is kept in-tact.

> Another difference is that in my implementation, when an event listener decide about the `AuthorizationRequest` (allowing or denying it), the response is cleared so the user is redirected only if there is no decision.

Could you please point me to the lines where you clear the response? The method signature is intentionally strict `setResponse(Response $response)`, to disallow null response. Anybody willing to change the decision about whether the response should be set or not should do something about it in a listener that set the response in the first place. Listeners should honour the decision of the listeners that have already been executed on the request.

I would argue that allowing other developers to disapprove the request once it has been approved from the listener is not nice. You should override the listener that gave the approval, or use some of its extension point to augment the decision process. Hence the `approveAuthorization()`.

At the end of the day, we could do all of this in the AuthorizationController, but we break it down to give each class it's single responsibility and allow for greater extensibility and all that....

> 
> I do not like the `authorization_strategy` related changes for different reasons:
> 
> * This functionality can be implemented without introducing another level of indirection. Simply configure the appropriate event listeners.
That level of indirection is the extension point for a well-defined part of the OAuth spec. The built-in delegating listener makes it clear that there is a step that handles authorization decision in the OAuth2 process, but does not care about the way it's handled. It just tells that there is this authorization task that must be fulfilled by `AuthorizationDecisionStrategy`. It's possible to make it simpler by changing the strategies into separate listeners and switch the listeners based on configuration but then you could not force the `AuthorizationDecisionStrategy` interface. 

> * To implement a new strategy, the bundle configuration should be modified, so it is difficult to extend it. If we only use event listeners, this problem is avoided.
That's just my oversight, can be fixed easily.

> * If I choose the `user_consent` strategy, I'd like to set the `$consentApprovalRoute` from the bundle configuration, so the bundle config gets more complicated.
True that.

> My thought was to left to the bundle user to decide how to implement the authorization process, but I think it can be convenient to provide an `AlwaysAllowAuthorizationRequestListener`, that could be easily enabled from a bundle configuration key, so we provide a minimal working implementation.

I want to allow anybody to design their own Authorization controller/view - but keep the rest on us. The issue with `AlwaysAllowAuthorizationRequestListener` is that you need to remove it from the container so that it doesn't run. Whereas in my implementation, you switch the strategy to your custom class that implements the interface and you are good to go. You don't have to understand the bundle to the level that listeners are the mechanism that augment the authorization request.


I'll push changes that hopefully fix most of your concerns, but I'd like to keep the strategy thing. Open to renaming it though.

File: Event/Listener/AuthorizationRequestAuthenticationListener.php (hunk 1 above)
Comment (MichaelKubovic) [reply Event/Listener/AuthorizationRequestAuthenticationListener.php]: added configuration option

File: Event/Listener/AuthorizationRequestAuthenticationListener.php (hunk 2 above)
Comment (MichaelKubovic) [reply Event/Listener/AuthorizationRequestAuthenticationListener.php]: done, current firewall is retrieved from `FirewallMap`

File: Event/Listener/AuthorizationRequestAuthenticationListener.php (hunk 3 above)
Comment (MichaelKubovic) [reply Event/Listener/AuthorizationRequestAuthenticationListener.php:78]: It does not really make sense to require IS_AUTHENTICATED_FULLY during OAuth authorization flow. That would require the user to enter his credentials any time he comes to authorization endpoint (this goes against the single sign-on I'm trying to build). However, if there will be a real requirement driven by the use-case, we can still make it into a configuration option later.

File: Event/AuthorizationRequestResolveEvent.php (hunk 4 above)
Comment (MichaelKubovic) [reply Event/AuthorizationRequestResolveEvent.php]: sure, my bad, fixed

File: Event/AuthorizationRequestResolveEvent.php (hunk 5 above)
Comment (MichaelKubovic) [reply Event/AuthorizationRequestResolveEvent.php]: my bad, fixed

File: Model/AuthorizationDecision/UserConsentDecisionStrategy.php (hunk 6 above)
Comment (MichaelKubovic) [reply Model/AuthorizationDecision/UserConsentDecisionStrategy.php]: configuration option added

File: DependencyInjection/Configuration.php (hunk 7 above)
Comment (MichaelKubovic) [reply DependencyInjection/Configuration.php]: I've changed it so that `authorization_strategy` accepts service id.

File: Event/Listener/AuthorizationRequestAuthenticationListener.php (hunk 3 above)
Comment (ajgarlag) [reply Event/Listener/AuthorizationRequestAuthenticationListener.php:78]: It makes sense for me if I store the client authorization to a persistent storage, so the user only have to consent once for each client (if scopes do not change). In that case, I want to be sure that the user is fully authenticated. I think GitHub authorization workflow works in a similar way.

File: Event/Listener/AuthorizationRequestAuthenticationListener.php (hunk 1 above)
Comment (ajgarlag) [reply Event/Listener/AuthorizationRequestAuthenticationListener.php]: Where is this configuration option?

**Comment by ajgarlag** [comment]
IMO there should be three ways to resolve an authorization request:
  1. Approve it
  2. Deny it (not allowed by your implementation)
  3. Redirect the user to a consent endpoint (setting a response)

With your implementation, if a listener set a response to the event, the user will receive that response; no matter if another listener approves the authorization later. And if a listener approves the authorization, another listener cannot deny it.

I prefer my `AuthorizationRequestResolveEvent::resolveAuthorization(bool $authorizationResolution)` which can allow or deny the authorization request, and clear the response if any; over your `AuthorizationRequestResolveEvent::approveAuthorization`. I think it's more flexible. If you stink think that the decision made from a listener should not be overriden from another one, you can call `Event::stopPropagation` from that listener.

**Comment by MichaelKubovic** [comment]
> If you stink think that the decision made from a listener should not be overriden from another one, you can call `Event::stopPropagation` from that listener.

Would that skip all the other built-in listeners that you actually want to run?

File: Event/AuthorizationRequestResolveEvent.php [hunk 8]
```diff
@@ -9,49 +9,36 @@
...
 
-    public function getAuhorizationResolution(): bool
-    {
-        return $this->authorizationResolution;
-    }
-
-    public function resolveAuthorization(bool $authorizationResolution): void
-    {
-        $this->authorizationResolution = $authorizationResolution;
-        $this->response = null;
```
Comment (ajgarlag) [inline Event/AuthorizationRequestResolveEvent.php]: Here, if a listener approves or denies the authorization request, the response is unset.

File: Event/AuthorizationRequestResolveEvent.php [hunk 9]
```diff
@@ -9,49 +9,36 @@
...
-    }
-
     public function hasResponse(): bool
     {
-        return $this->response instanceof ResponseInterface;
+        return $this->response !== null;
```
Comment (ajgarlag) [inline Event/AuthorizationRequestResolveEvent.php:38]: Why this change?

File: Event/AuthorizationRequestResolveEvent.php [hunk 10]
```diff
@@ -9,49 +9,36 @@
...
     }
 
-    public function getResponse(): ResponseInterface
+    public function getResponse(): ?ResponseInterface
```
Comment (ajgarlag) [inline Event/AuthorizationRequestResolveEvent.php:41]: This method wont return null.

**Comment by ajgarlag** [comment]
> > If you stink think that the decision made from a listener should not be overriden from another one, you can call `Event::stopPropagation` from that listener.
> 
> Would that skip all the other built-in listeners that you actually want to run?

Yes, so I prefer to allow any listener to override a previous decision. But with your implementation, if a listener has set a response, another one cannot approve the request later, so I think in that case, you should make it explicit calling `Event::stopPropagation`.

With my `AuthorizationRequestResolveEvent::resolveAuthorization(bool $authorizationResolution)` we avoid that problem. But maybe it's not clear that the authorization request can be resolved calling to `resolveAuthorization` or `setResponse`.

What do you think about adding a `AuthorizationRequestResolveEvent::resolve(bool|ResponseInterface $resolution)` method and removing the `AuthorizationRequestResolveEvent::setResponse` method. With that implementation, the last call to the `resolve` method would win the authorization resolution.

**Comment by MichaelKubovic** [comment]
Right, I prefer not to run listener that has no impact (because its outcome is supersede by the succeeding listener). It seems more natural to change the way the authorization is resolved, to fix the decision process that produced the original resolution in the first place. When you just add another listener that changes previous decision - what is the point in running the original listener at all?

I see some analogy with Gaurd authenticators. When some of them set the response, they are done.
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php#L91

**Comment by MichaelKubovic** [comment]
A use-case for the listener that overrides previously generated response or authorization resolution  would help me understand your motivation.
